### PR TITLE
tracing: internally remove WithSeparateRecording

### PR DESCRIPTION
**Pull request by tbg**
- tracing: improve a comment
- tracing: internally remove WithSeparateRecording
- tracing: don't erroneously return a noopSpan
- tracing: return nil *SpanMeta

### CODE REVIEW

**Comment by tbg** [comment]
@asubiotto I had a little lightbulb moment following our conversation earlier today. We don't need the "separate recording" option, really what we want is to treat the span in the processor as behind a "forced RPC boundary". This is exactly what you get if you use `WithRemoteParent` instead of `WithParent`. So we get to switch to that, remove the `WithSeparateRecording` option, and everything works just the same (mod perhaps a small perf degradation, but I'll polish perf at some point to make sure it's not noticeable anyway, definitely nothing there that couldn't be optimized away).

**Review by asubiotto — DISMISSED** [review]
> Reviewed 1 of 1 files at r1, 6 of 6 files at r2, 2 of 2 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56343)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, and @tbg)

___
*[pkg/sql/colflow/vectorized_flow.go, line 946 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56343#-MLRqQRv-rWXEBAkVRKo:-MLRqQRv-rWXEBAkVRKp:b-177icb) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce4f901d4fdb338ee47e62f8fbe553cfc18f61d/pkg/sql/colflow/vectorized_flow.go#L946)):*
> ```Go
> 						// Start a separate recording so that GetRecording will return
> 						// the recordings for only the child spans containing stats.
> 						ctx, span := tracing.ChildSpanRemote(ctx, "")
> ```

The confusing thing to me is that this might happen on the gateway node (i.e. this call is not always "remote"). What about deriving a "remote" child in `SetupFlow` (look at `parentSpan`), is that attribute inherited? In that case, we could just call `ChildSpan` here and it'll implicitly be remote or not depending on the ancestor.

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56343)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @irfansharif, and @knz)

___
*[pkg/sql/colflow/vectorized_flow.go, line 946 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56343#-MLRqQRv-rWXEBAkVRKo:-MLlC43pAv9LGaKPXDpF:b-91h49m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce4f901d4fdb338ee47e62f8fbe553cfc18f61d/pkg/sql/colflow/vectorized_flow.go#L946)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

The confusing thing to me is that this might happen on the gateway node (i.e. this call is not always "remote"). What about deriving a "remote" child in `SetupFlow` (look at `parentSpan`), is that attribute inherited? In that case, we could just call `ChildSpan` here and it'll implicitly be remote or not depending on the ancestor.
</blockquote></details>

I don't think the current API explains things well enough to make sense.
There is no "judgement" here about whether the parent span is truly remote. Instead, what we are doing is using the `.Meta()` of the parent span to create the child (`WithRemoteParent(sp.Meta())`) instead of hooking them up directly (`WithParent(sp)`).

This business of interconnecting spans directly locally (`WithParent`) is the crux in why our semantics are more complicated compared to a "true" distributed tracer. In CRDB, "most of the time" you use `WithParent`, as that is the easiest thing to do to get recordings where they're needed, but you can opt out of that (perhaps because you have to, see RPC boundary) and use `WithRemoteParent` and take up an obligation to make sure that the recording actually does get handed to the parent, eventually.

In the case at hand, we always do hand the recording to the parent span via the trailingMeta, and so we're opting out intentionally.

I sense that the confusion stems around the `Remote` in `WithRemoteParent`. I wonder what else we could name these concepts so that they are more intuitive here and not tied to the local/remote nomenclature. Ultimately it's about whether the trace data will be included "automatically" (because it's tied to a local subtree of the trace that already exists) or whether the caller has to propagate it. Perhaps instead of `WithRemoteParent`, it's `WithManualParent`, as opposed to an "automatic" parent? I guess this name isn't self-explanatory, but it's clear that `WithManualParent` means there is some extra work involved, and it doesn't give you wrong ideas about when it's appropriate to use. What do you think?

**Comment by knz** [comment]
WithParentAndAutoCollection / WithParentAndExplicitCollection

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56343)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

___
*[pkg/sql/colflow/vectorized_flow.go, line 946 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56343#-MLRqQRv-rWXEBAkVRKo:-MLm9HWl4kohM45nmCqJ:b8aju10) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce4f901d4fdb338ee47e62f8fbe553cfc18f61d/pkg/sql/colflow/vectorized_flow.go#L946)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I don't think the current API explains things well enough to make sense.
There is no "judgement" here about whether the parent span is truly remote. Instead, what we are doing is using the `.Meta()` of the parent span to create the child (`WithRemoteParent(sp.Meta())`) instead of hooking them up directly (`WithParent(sp)`).

This business of interconnecting spans directly locally (`WithParent`) is the crux in why our semantics are more complicated compared to a "true" distributed tracer. In CRDB, "most of the time" you use `WithParent`, as that is the easiest thing to do to get recordings where they're needed, but you can opt out of that (perhaps because you have to, see RPC boundary) and use `WithRemoteParent` and take up an obligation to make sure that the recording actually does get handed to the parent, eventually.

In the case at hand, we always do hand the recording to the parent span via the trailingMeta, and so we're opting out intentionally.

I sense that the confusion stems around the `Remote` in `WithRemoteParent`. I wonder what else we could name these concepts so that they are more intuitive here and not tied to the local/remote nomenclature. Ultimately it's about whether the trace data will be included "automatically" (because it's tied to a local subtree of the trace that already exists) or whether the caller has to propagate it. Perhaps instead of `WithRemoteParent`, it's `WithManualParent`, as opposed to an "automatic" parent? I guess this name isn't self-explanatory, but it's clear that `WithManualParent` means there is some extra work involved, and it doesn't give you wrong ideas about when it's appropriate to use. What do you think?
</blockquote></details>

I like @knz's naming proposal that indicates where the collection/propagation responsibility lies. I think this is clearer to me than the "remote" term and feels more generally applicable.

**Review by irfansharif — APPROVED** [review]
The simplification here makes sense. I'll leave y'all to duke it out over the naming, though I do like Rafa's suggestion talking in terms of recording collection.

**Review by knz — APPROVED** [review]
LGTM modulo the naming

> Reviewed 3 of 6 files at r2, 1 of 2 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56343#-:-MLqroMSAuTwft4p7U4c:b-klvgdi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

**Comment by tbg** [comment]
Changed the names, TFTR!

bors r=irfansharif,knz

**Comment by tbg** [comment]
`ThreadSanitizer: clock allocator overflow (65536*1024). Dying.` in sql tests

bors r=irfansharif,knz

**Comment by tbg** [comment]
bors r-
Just in case it's me

**Comment by tbg** [comment]
bors single on
bors r=irfansharif,knz

I saw the same problem on https://github.com/cockroachdb/cockroach/issues/55465 (again on the sql package), so I have convinced myself that this has nothing to do with my change. I hit it before because my change was always lumped in with something that changed `./pkg/sql` which triggered that package to be stressraced.

**Comment by knz** [comment]
This is an error in the go runtime. No relationship to this PR indeed.

Two general directions to alleviate:

- make it so that there is less code executed in a package's full test run. Here the `sql` package is simply too large.

- run `testrace` with fewer cores.

- don't run `testrace` on all the tests. Note that we already have logic for `teamcity-testrace` to only run it on packages directly affected by the PR. Your PR is touching `sql` so that's how you get in this situation. Maybe we can exclude `sql` specifically from `teamcity-testrace`?

That being said, I suspect that not just your PR is affected. I would expect that every SQL PR will be blocked on a similar situation at this point.

**Comment by tbg** [comment]
I am getting the clock allocator overflow much more reliably on this PR. I started four testrace runs off this PR, and three of them hit the error (the fourth one passed):

https://teamcity.cockroachdb.com/viewLog.html?buildId=2441374&buildTypeId=Cockroach_UnitTests_Testrace
https://teamcity.cockroachdb.com/viewLog.html?buildId=2441376&buildTypeId=Cockroach_UnitTests_Testrace
https://teamcity.cockroachdb.com/viewLog.html?buildId=2441378&buildTypeId=Cockroach_UnitTests_Testrace

I did the same on a PR that touched sql just to trigger the same testrace run, and all four passed:

https://github.com/cockroachdb/cockroach/pull/56646
https://teamcity.cockroachdb.com/viewType.html?buildTypeId=Cockroach_UnitTests_Testrace&branch_Cockroach_UnitTests=56646&tab=buildTypeStatusDiv

Now that I'm relatively convinced that I am making something worse in this PR - it does make some sense, I am touching tracing, and in particular one of the commits removes an "optimization" that erroneously avoided creating spans - maybe the extra spans that are being created now are pushing us over the edge, somehow.

Will look into this more next week.

**Comment by tbg** [comment]
Got it, see the last commit. PTAL

**Review by irfansharif — APPROVED** [review]

**Comment by tbg** [comment]
Thanks Irfan.

bors r=irfansharif,knz,asubiotto

**Comment by tbg** [comment]
For posterity, I triggered the testrace build 5x on this PR and they all went through. Previously three out of four would fail.

**Comment by tbg** [comment]
bors r=irfansharif,knz,asubiotto

**Pull request merged**
